core/src/main/java/org/bitcoinj/core/Coin.java

```diff
@@ -118,6 +118,36 @@ public static Coin valueOf(final int coins, final int cents) {
         return coin;
     }
 
+    /**
+     * Convert a decimal amount of bitcoin into satoshis.
+     *
+     * @param coins number of coins
+     * @return number of satoshis
+     */
+    public static long bitcoinToSatoshi(BigDecimal coins) {
```

Comment:
Hmm, do we need this and the counterpart method as _public_ methods? Actually the idea of the `Coin` class is that people don't fiddle around with monetary amounts themselves, so methods like this should not be needed.

Also I'm not so happy about the use of the word bitcoin in the method names – satoshis are Bitcoin too. Perhaps BTC would be more fitting, as that implies the denomination?

Reply:
> do we need this and the counterpart method as _public_ methods?

I thought about keeping them `private`, but ultimately decided they were useful as `public`.

> Actually the idea of the `Coin` class is that people don't fiddle around with monetary amounts themselves, so methods like this should not be needed.

 I view **bitcoinj** as the reference implementation of Bitcoin constants and types for the JVM ecosystem. There are clearly cases where people using **bitcoinj** will have `long` values of satoshi they need to convert to (big)decimal values of bitcoin. I think there should be a reference conversion function and that it should live in the `Coin` class.

There are use cases where people will be developing bitcoin apps where they're going to need to use a charting library that takes BigDecimal (or maybe float, ugh). They want to use a math library that works with BigDecimal, etc. They will need to convert back and forth to `Coin` and even to just `long` when they interact with transactions, etc.

In ConsensusJ I ended up writing this conversion code twice (because I didn't want the two libraries to have dependencies on each other.) And I didn't write the code as well as you and Mike Hearn did! I just discovered the shift point functions this weekend looking at `Coin.java`!

So it's nice to have the factory method and the `toBitcoin` instance method, but I also think bitcoinj clients will want to be able to convert `long` to `BigDecimal` and vice-versa without unnecessary object creation.


> 
> Also I'm not so happy about the use of the word bitcoin in the method names – satoshis are Bitcoin too. Perhaps BTC would be more fitting, as that implies the denomination?

I'd be ok with BTC, but (lower case) bitcoin is also a [denomination](https://en.bitcoin.it/wiki/Units#Table_of_all_units).

Reply:
>the idea of the Coin class is that people don't fiddle around with monetary amounts themselves, so methods like this should not be needed.

I agree that we want to discourage users from doing this. Maybe I should add some text to the class-level JavaDoc comment talking about how to use `Coin`, to avoid using floating-point and to use `long` and `BigDecimal` when you absolutely need a primitive-ish type to interact with other libraries.

Reply:
What do you think about extending the Coin API to allow for this?

- BigDecimal→long: `long sat = Coin.ofBtc(btc).toSat()`
- long→BigDecimal: `BigDecimal btc = Coin.ofSat(sat).toBtc()`

Reply:
> What do you think about extending the Coin API to allow for this?
> 
> * BigDecimal→long: `long sat = Coin.ofBtc(btc).toSat()`
> * long→BigDecimal: `BigDecimal btc = Coin.ofSat(sat).toBtc()`

This looks pretty good. But for high-performance or large numbers of conversions there is an object creation and disposal that could be avoided with pure `static` functions.

Reply:
I renamed `Bitcoin` to `Btc` in the appropriate places and added the `ofSat` and `toSat` methods (don't know if you want to deprecate the duplicate `valueOf` and `longValue` methods)